Source/WebCore/platform/graphics/skia/SkiaThreadedPaintingPool.cpp

```diff
@@ -0,0 +1,268 @@
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. ``AS IS'' AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL APPLE INC. OR
+ * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+ * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "SkiaThreadedPaintingPool.h"
+
+#if USE(COORDINATED_GRAPHICS) && USE(SKIA)
+#include "GLContext.h"
+#include "PlatformDisplay.h"
+#include "ProcessCapabilities.h"
+
+#include <wtf/NumberOfCores.h>
+#include <wtf/text/StringToIntegerConversion.h>
+
+constexpr bool dumpPaintTaskSettings = true;
+constexpr bool collectPaintTaskStatistics = false;
+constexpr Seconds dumpPaintTaskStatisticsInterval = 5_s;
+
+static unsigned long s_numberOfPaintTasks = 0;
+static unsigned long s_numberOfPaintTasksDispatchedToCPU = 0;
+static unsigned long s_numberOfPaintTasksDispatchedToGPU = 0;
+static unsigned long s_numberOfPaintTasksAboveGPUSizeThreshold = 0;
+
+namespace WebCore {
+
+class SkiaThreadedPaintingPool::Worker : public AutomaticThread {
+protected:
+    friend class SkiaThreadedPaintingPool;
+
+    Worker(const AbstractLocker& locker, SkiaThreadedPaintingPool& pool, Box<Lock> lock, Ref<AutomaticThreadCondition>&& condition)
+        : AutomaticThread(locker, lock, WTFMove(condition), ThreadType::Graphics, Seconds::infinity())
+        , m_pool(pool)
+    {
+    }
+
+    bool shouldSleep(const AbstractLocker&) final { return false; }
+    ASCIILiteral name() const final { return "SkiaThreadedPaintingPool"_s; }
+
+    PollResult poll(const AbstractLocker&) final
+    {
+        if (m_pool.m_tasks.isEmpty())
+            return PollResult::Wait;
+
+        m_task = m_pool.m_tasks.takeFirst();
+        if (!m_task.task)
+            return PollResult::Stop;
+
+        return PollResult::Work;
+    }
+
+    WorkResult work() final
+    {
+        ASSERT(m_task.task);
+
+        bool isAccelerated = m_task.buffer->isBackedByOpenGL();
+
+        {
+            Locker locker { *m_pool.m_lock };
```

Comment:
I can get rid of the locking in `work()` -- it was only used to track the number of active CPU/GPU tasks, which we no longer use in the scheduling logic.